File: gcj2013QRA.py
#!/usr/bin/env python
import sys

def solve():
    xs = [sys.stdin.readline().rstrip() for _ in range(5)][:4]
    def won(c):
        def line(w): return all([x in [c,'T'] for x in w])
        def lines(xs): return any(map(line,xs))
        def cross(xs):
            return [[xs[i][j] for i,j in enumerate(n)]
                    for n in [range(4),reversed(range(4))]]
        return any(map(lines,[xs,zip(*xs),cross(xs)]))
    if won('X'): return "X won"
    if won('O'): return "O won"
    if any(['.' in w for w in xs]): return "Game has not completed"
    return "Draw"

for x in range(int(sys.stdin.readline())):
    print("Case #"+str(x+1)+": "+str(solve()))

File: gcj2013QRA.hs
#!/usr/bin/env runghc
import Data.List
import Data.List.Split
main = getContents >>= mapM_ putStrLn . zipWith caseno [1..] . solve . lines
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map answer . take (read c) . analyze $ cs ::[String]
analyze = map (take 4) . chunksOf 5
answer xs
  | won 'X' xs = "X won"
  | won 'O' xs = "O won"
  | any ('.' `elem`) xs = "Game has not completed"
  | otherwise = "Draw"
won c xs = any (all (`elem` [c,'T'])) $ concatMap ($xs) [id,transpose,cross]
cross xs = map (zipWith (!!) xs) [[0..],[3,2..]]
